dapp/src/components/buySell/ApproveSwap.js

```diff
@@ -0,0 +1,282 @@
+import React, { useState, useEffect } from 'react'
+import { fbt } from 'fbt-runtime'
+import { useStoreState } from 'pullstate'
+import { useWeb3React } from '@web3-react/core'
+import ContractStore from 'stores/ContractStore'
+import analytics from 'utils/analytics'
+import withRpcProvider from 'hoc/withRpcProvider'
+import { ethers } from 'ethers'
+import withIsMobile from 'hoc/withIsMobile'
+
+const ApproveSwap = ({
+  stableCoinToApprove,
+  needsApproval,
+  selectedSwap,
+  swapMetadata,
+  onSwap,
+  allowancesLoaded,
+  onMintingError,
+  formHasErrors,
+  swappingGloballyDisabled,
+  storeTransaction,
+  storeTransactionError,
+  rpcProvider,
+  isMobile,
+}) => {
+  const [stage, setStage] = useState('approve')
+  const [contract, setContract] = useState(null)
+  const [isApproving, setIsApproving] = useState({})
+  const [show, setShow] = useState(false)
+  const web3react = useWeb3React()
+  const { library, account } = web3react
+  const coinApproved = stage === 'done'
+
+  const approvalNeeded =
+    ((!selectedSwap ||
+      formHasErrors ||
+      swappingGloballyDisabled ||
+      !allowancesLoaded ||
+      !needsApproval) &&
+      !coinApproved) ||
+    !show
+
+  useEffect(() => {
+    setShow(false)
```

Comment:
this stuff hides the approval button right after switching currencies, otherwise the wrong currency is shown on the button for a split second

Reply:
Aha I see thanks for explaining. Made a small change in interest to remove the `show` flag: 
https://github.com/OriginProtocol/origin-dollar/pull/958/commits/a5eeaab5059461f6aba1d619bbe872e10e830860

Moved the global setter of `swapEstimations` out of the `setTimeout` function in case when delay is 0. 

While your approach definitely worked (arguably worked better since in some cases the wrong stablecoin would still be shown for 1 render frame) I really like to have as minimal amount of moving parts (show flag in this case) that can control the rendering of the component.

I realise this might seem like a weird obsession (and it did to me as well in the past), but working on the same codebase for years really makes you appreciate when things are simple. Sometimes code complexity can really grow so big (and I have been guilty of that in the past) that development grinds to a halt, because there are so many interconnected moving parts that making one change breaks 3 other things.